orangecontrib/timeseries/widgets/owspiralogram.py

```diff
@@ -277,170 +462,609 @@ class Inputs:
+            labels = [str(i + off) for i in range(self.nperiods)]
+        items = []
+        for label in labels:
+            item = QGraphicsSimpleTextItem(label)
+            item.setFont(font)
+            items.append(item)
+        return items
+
+    def draw_legend(self):
+        legend = self.legend
+        scene_rect = self.scene.itemsBoundingRect()
+        legend_rect = legend.boundingRect()
+        # PaletteItemSample adds 20 to bounding rect height
+        legend.setPos(scene_rect.right() + 10,
+                      scene_rect.bottom() - legend_rect.height()
+                      + 20 * isinstance(legend, PaletteItemSample))
+        self.scene.addItem(legend)
+
+    # Selection
+    # ---------
+
+    def _on_segment_clicked(self, segment, event):
+        if event.button() != Qt.LeftButton:
+            event.ignore()
+            return
+        x, r = segment.x, segment.r
+        if event.modifiers() & Qt.ShiftModifier \
+                and self.last_selected in self.selection:
+            xlast, rlast = self.last_selected
+            target = {
+                (sx % self.nperiods, sr)
+                for sr in range(min(r, rlast), max(r, rlast) + 1)
+                for sx in range(xlast - self.nperiods * (x < xlast), x + 1)
+            }
+        else:
+            target = {(x, r)}
+            self.last_selected = (x, r)
+
+        if event.modifiers() & Qt.ControlModifier:
+            flag = Select
+        elif self.selection == target:
+            flag = Clear
+        else:
+            flag = ClearAndSelect
+
+        self.select(target, flag)
+
+    def _on_scene_clicked(self):
+        self.select(None, Clear)
+
+    def select(self, selection, flag):
+        if flag == Clear:
+            to_update = self.selection
+            self.selection = set()
+        elif flag == ClearAndSelect:
+            to_update = self.selection ^ selection
+            self.selection = selection
+        else:
+            assert flag == Select
+            to_update = selection
+            self.selection |= selection
+
+        for coord in to_update:
+            if coord in self.segments:
+                self.segments[coord].set_selected(coord in self.selection)
+        self.commit_selection()
+
+    # Commits
+    # -------
+
+    def commit_statistics(self):
+        self.Outputs.statistics.send(self.computed_data)
+
+    def commit_selection(self):
+        if not self.selection:
+            data = None
+        else:
+            seg_indices = self.block_data.indices
+            rows = np.hstack([seg_indices[coord] for coord in self.selection])
+            data = self.data[rows]
+        self.Outputs.selected_data.send(data)
 
 
 if __name__ == "__main__":
-    WidgetPreview(OWSpiralogram).run(Table.from_file('airpassengers'))
+    WidgetPreview(OWSpiralogram).run(
+        Timeseries.from_file(
+            '/Users/janez/Downloads/slovenia-traffic-accidents-2016-events.tab')
```

Comment:
I am sorry, but this will not pass my review. 😉

Reply:
Well, I tried. :)